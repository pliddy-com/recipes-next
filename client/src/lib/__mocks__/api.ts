export interface ApiTestProps {
  getNavTaxonomy: () => [{ slug: string }];
  getRecipeSlugs: () => string[];
  getTagSlugs: () => string[];
  getRecipeIndex: () => [{ slug: string }];
  getRecipeList: () => [{ slug: string }];
  getRecipePage: () => { slug: string; title: string };
  getTagIndex: () => [{ slug: string; title: string }];
}

jest.mock('urql');

// import api library to mock
const api: ApiTestProps = jest.createMockFromModule('lib/api');

api.getNavTaxonomy = jest.fn().mockResolvedValue({
  categories: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ],
  cuisine: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ],
  tags: [
    { slug: 'slug-1', title: 'Title 1' },
    { slug: 'slug-2', title: 'Title 2' }
  ]
});

api.getRecipeSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getTagSlugs = jest.fn().mockResolvedValue(['slug-1', 'slug-2']);

api.getRecipeIndex = jest.fn().mockResolvedValue([
  {
    slug: 'slug-1',
    title: 'Title 1'
  },
  {
    slug: 'slug-2',
    title: 'Title 2'
  }
]);

api.getRecipeList = jest.fn().mockResolvedValue([
  {
    __typename: 'Tag',
    slug: 'baking',
    title: 'Baking',
    linkedFrom: {
      recipeCollection: {
        total: 3,
        items: [
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-06-17T00:02:14.528Z',
              __typename: 'Sys'
            },
            title: 'Biscuits',
            slug: 'biscuits',
            abstract:
              'Traditional baking powder biscuits that use grated frozen butter and folded dough to give the biscuits a soft, flaky texture.',
            image: {
              __typename: 'Asset',
              title: 'Biscuits',
              description:
                'A batch of baking soda biscuits on a parchment-lined baking sheet.',
              contentType: 'image/jpeg',
              fileName: 'biscuits.jpg',
              size: 2583242,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/B6C4D23gXNDNfGVrxb1DK/9d8e62c1faf660f7e542eed14e044b57/biscuits.jpg',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Biscuits',
                  slug: 'biscuits'
                },
                {
                  __typename: 'Tag',
                  title: 'Breakfast',
                  slug: 'breakfast'
                },
                {
                  __typename: 'Tag',
                  title: 'Southern',
                  slug: 'southern'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          },
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-11-15T21:50:41.051Z',
              __typename: 'Sys'
            },
            title: 'No-Knead Olive Bread',
            slug: 'no-knead-olive-bread',
            abstract:
              'A savory variation of no-knead bread with the addition of olives, garlic, and rosemary to give it a Mediterranean twist.',
            image: {
              __typename: 'Asset',
              title: 'Olive Bread',
              description: 'Loaf of no-knead olive bread',
              contentType: 'image/jpeg',
              fileName: 'olive-bread.JPG',
              size: 2849709,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/2rUa4IhfkLtRRBmHS6ihcD/e953337e00d1eba9551b15e7eee18229/olive-bread.JPG',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Bread',
                  slug: 'bread'
                },
                {
                  __typename: 'Tag',
                  title: 'Olives',
                  slug: 'olives'
                },
                {
                  __typename: 'Tag',
                  title: 'Mediterranean',
                  slug: 'mediterranean'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          },
          {
            __typename: 'Recipe',
            sys: {
              firstPublishedAt: '2022-06-18T22:42:01.107Z',
              __typename: 'Sys'
            },
            title: 'No-Knead Bread',
            slug: 'no-knead-bread',
            abstract:
              'No-knead bread using instant yeast and a dutch oven or covered ceramic bread pan for a classic, crispy crust.',
            image: {
              __typename: 'Asset',
              title: 'No Knead Bread',
              description:
                'Loaf of no-knead bread cooling on a wire baking rack.',
              contentType: 'image/jpeg',
              fileName: 'no-knead-bread.JPG',
              size: 2943075,
              url: 'https://images.ctfassets.net/fo9qwg6zarbt/6HZDSm8AK4iQHr9v2UOGpq/17569d5be00441216338c7800e5ebc2c/no-knead-bread.JPG',
              height: 3024,
              width: 4032
            },
            tagsCollection: {
              items: [
                {
                  __typename: 'Tag',
                  title: 'Baking',
                  slug: 'baking'
                },
                {
                  __typename: 'Tag',
                  title: 'Bread',
                  slug: 'bread'
                }
              ],
              __typename: 'RecipeTagsCollection'
            }
          }
        ],
        __typename: 'RecipeCollection'
      },
      __typename: 'TagLinkingCollections'
    }
  }
]);

api.getRecipePage = jest.fn().mockResolvedValue({
  recipe: {
    title: 'Recipe Title',
    slug: 'slug-1',
    __typename: 'Recipe',
    sys: { firstPublishedAt: '2022-06-26T07:24:30.384Z' },
    abstract: 'abstract',
    recipeYield: 4,
    prepTime: 15,
    cookTime: 60,
    keywords: ['keyword1', 'keyword2'],
    image: {
      sys: {
        id: 'sysid-1',
        __typename: 'Sys'
      },
      __typename: 'Asset',
      title: 'Image Title',
      description: 'Image description.',
      contentType: 'image/jpeg',
      fileName: 'filename.jpg',
      size: 99999,
      url: 'https://recipes.pliddy.com/url',
      height: 300,
      width: 400
    },
    ingredientsList: [
      {
        sectionTitle: 'Section Label 1',
        sectionItems: ['ingredient 1', 'ingredient 2']
      }
    ],
    equipment: ['equipment 1', 'equipment 2'],
    instructionsList: [
      {
        sectionTitle: 'Section Label 1',
        sectionItems: ['Instructions item 1']
      },
      {
        sectionTitle: 'Section Label 2',
        sectionItems: ['Instructions item 2', 'Instructions item 3']
      }
    ],
    notes: ['note 1'],
    tagsCollection: {
      items: [
        {
          sys: {
            id: 'sysid-4',
            __typename: 'Sys'
          },
          __typename: 'Tag',
          title: 'Tag 1',
          slug: 'tag-1'
        }
      ],
      __typename: 'RecipeTagsCollection'
    }
  },
  cuisine: [
    {
      __typename: 'Tag',
      title: 'Tag 1 Title',
      slug: 'tag-1',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-1',
              title: 'Title 2'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-2',
              title: 'Title 1'
            }
          ]
        }
      }
    },
    {
      __typename: 'Tag',
      title: 'Tag 2 Title',
      slug: 'tag-2',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-3',
              title: 'Title 3'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-4',
              title: 'Title 5'
            }
          ]
        }
      }
    }
  ],
  categories: [
    {
      __typename: 'Tag',
      title: 'Tag 1 Title',
      slug: 'tag-1',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-1',
              title: 'Title 2'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-2',
              title: 'Title 1'
            }
          ]
        }
      }
    },
    {
      __typename: 'Tag',
      title: 'Tag 2 Title',
      slug: 'tag-2',
      linkedFrom: {
        recipeCollection: {
          total: 2,
          items: [
            {
              __typename: 'Recipe',
              slug: 'slug-3',
              title: 'Title 3'
            },
            {
              __typename: 'Recipe',
              slug: 'slug-4',
              title: 'Title 5'
            }
          ]
        }
      }
    }
  ]
});

api.getTagIndex = jest.fn().mockResolvedValue([
  {
    __typename: 'Tag',
    title: 'Tag 1 Title',
    slug: 'tag-1',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-1',
            title: 'Title 2'
          },
          {
            __typename: 'Recipe',
            slug: 'slug-2',
            title: 'Title 1'
          }
        ]
      }
    }
  },
  {
    __typename: 'Tag',
    title: 'Tag 2 Title',
    slug: 'tag-2',
    linkedFrom: {
      recipeCollection: {
        total: 2,
        items: [
          {
            __typename: 'Recipe',
            slug: 'slug-3',
            title: 'Title 3'
          },
          {
            __typename: 'Recipe',
            slug: 'slug-4',
            title: 'Title 5'
          }
        ]
      }
    }
  }
]);

module.exports = api;
