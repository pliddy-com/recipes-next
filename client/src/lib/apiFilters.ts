import {
  TagDefaultFragment,
  NavTaxonomyFragment,
  TagTaxonomyFragment,
  RecipeSlugFragment
} from '../types/queries';

import { hasValue } from './utils';

export const filterSlugs = (
  items: (RecipeSlugFragment | TagDefaultFragment | null)[]
) =>
  items &&
  items
    ?.map((child) => (child && child.slug ? child.slug : null))
    .filter(hasValue);

export const filterTagsWithRecipes = ({
  tagCollection
}: {
  tagCollection: TagTaxonomyFragment;
}) =>
  tagCollection &&
  tagCollection?.items
    .filter(hasValue)
    .filter(
      (item) =>
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0
    );

export const filterTaxonomyItemsWithRecipes = ({
  taxonomyCollection
}: {
  taxonomyCollection: NavTaxonomyFragment;
}) => {
  return taxonomyCollection?.items?.[0]?.childrenCollection?.items?.filter(
    (item) =>
      item &&
      (('linkedFrom' in item &&
        item.linkedFrom &&
        'recipeCollection' in item.linkedFrom && // is a tag
        item?.linkedFrom?.recipeCollection?.total &&
        item?.linkedFrom?.recipeCollection?.total > 0) ||
        ('childrenCollection' in item && // is a taxonomy
          item?.childrenCollection?.total &&
          item?.childrenCollection?.total > 0))
  );
};

// TODO: update to filter out childrenCollection.items.linkedFrom.recipeCollection.total <== 0
