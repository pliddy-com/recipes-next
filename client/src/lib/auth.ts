/* istanbul ignore file */

import { AuthenticationDetails, CognitoUser } from 'amazon-cognito-identity-js';

import userPool from './userPool';

const APP_CLIENT_ID = process.env.NEXT_PUBLIC_COGNITO_APP_CLIENT_ID;
const USER_POOL_ID = process.env.NEXT_PUBLIC_COGNITO_USER_POOL_ID;
// const REGION = process.env.NEXT_PUBLIC_COGNITO_REGION;

interface signInProps {
  email: string;
  password: string;
}

export const signIn = async ({ email, password }: signInProps) => {
  if (!(email && password && APP_CLIENT_ID && USER_POOL_ID))
    throw new Error('Must provide email and password to sign in');
  try {
    // const poolData = {
    //   UserPoolId: USER_POOL_ID,
    //   ClientId: APP_CLIENT_ID
    // };

    // const userPool = new CognitoUserPool(poolData);

    const user = new CognitoUser({
      Username: email,
      Pool: userPool
    });

    const authDetails = new AuthenticationDetails({
      Username: email,
      Password: password
    });

    const res = user.authenticateUser(authDetails, {
      onSuccess: (result) => {
        console.log('login success', result);
      },
      onFailure: (err) => {
        console.log('login failure', err);
      },
      newPasswordRequired: (data) => {
        console.log('new password required', data);
      }
    });

    console.log({ res });
  } catch (e) {
    console.error(e);
    throw e;
  }
};
