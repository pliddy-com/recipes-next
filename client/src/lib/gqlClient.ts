/* istanbul ignore file */
import { cacheExchange, Client, fetchExchange, TypedDocumentNode } from 'urql';
import { ResultOf, VariablesOf } from '@graphql-typed-document-node/core';

const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_ACCESS_TOKEN;
const API_ENDPOINT = process.env.NEXT_PUBLIC_CONTENTFUL_CONTENT_API;

export const gqlClient = new Client({
  url: API_ENDPOINT || '',
  exchanges: [cacheExchange, fetchExchange],
  fetchOptions: () => ({
    headers: { authorization: ACCESS_TOKEN ? `Bearer ${ACCESS_TOKEN}` : '' }
  })
});

export async function queryGraphQLContent<
  TDoc extends TypedDocumentNode<ResultOf<TDoc>, VariablesOf<TDoc>>,
  TVars = TDoc extends unknown ? never : VariablesOf<TDoc>
>(document: TDoc, variables?: TVars) {
  const { data, error } = await gqlClient
    // @ts-expect-error query variables are typed by graphQLRequest
    .query(document, variables)
    .toPromise();

  if (!data) {
    if (error) {
      throw error;
    }
    throw new Error('No data returned');
  }

  return data;
}
