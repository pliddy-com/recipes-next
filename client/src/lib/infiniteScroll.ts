/**
 *  getPagedRecipes parsing function (parse in realtime instead of pre-rendering array of arrays?)
 */

export interface PaginateResultsProps {
  data: (object | null)[];
  pageSize: number;
}

export const paginateResults = ({ data, pageSize }: PaginateResultsProps) => {
  const numPages = Math.ceil(data.length / pageSize);
  const results = [];

  for (let pageNum = 0; pageNum < numPages; pageNum++) {
    const startItem = pageNum * pageSize;
    const pageData = data.slice(startItem, startItem + pageSize);
    results.push(pageData);
  }

  return results;
};

/**
 *  LoadNext function for infinite scrolling recipes
 */

export interface LoadNextProps {
  pagedResults: (object | null)[][];
  pageNum: number;
  pageSize: number;
  setDataCallback: (pagedResults: (object | null)[][]) => void;
  setPageNumCallback: (num: number) => void;
}

export const loadNext = ({
  pagedResults,
  pageNum,
  setDataCallback,
  setPageNumCallback
}: LoadNextProps): void => {
  if (pageNum < pagedResults.length) setPageNumCallback(pageNum + 1);
  setDataCallback(pagedResults.slice(0, pageNum + 1));
};
