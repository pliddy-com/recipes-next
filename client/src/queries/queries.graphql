# query for Taxonomy with Category & Subcategory levels
# getNavTaxonomy

query navMenuData {
  categories: taxonomyCollection(
    where: { slug: "categories" }
    order: title_ASC
    limit: 1
  ) {
    ...NavTaxonomy
  }

  cuisine: taxonomyCollection(
    where: { slug: "cuisine" }
    order: title_ASC
    limit: 1
  ) {
    ...NavTaxonomy
  }

  tags: tagCollection(order: title_ASC) {
    ...TagTaxonomy
  }
}

# Query for recipe slugs
# getRecipeSlugs

query recipeSlugs($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    __typename
    items {
      ...RecipeSlug
    }
  }
}

# query for category slugs
# getTagSlugs

query tagSlugs($where: TagFilter) {
  tagCollection(where: $where, order: title_ASC) {
    ...TagTaxonomy
  }
}

# query for lists of linked entries for tags and images
# getLinkedEntries

query linkedEntries {
  tags: tagCollection(order: title_ASC) {
    items {
      ...TagDefault
    }
  }
  images: assetCollection(order: fileName_ASC) {
    items {
      ...ImageDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)
# getRecipeList

query recipeList($where: TagFilter) {
  __typename
  tagCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    total
    items {
      ...ListPageItem
    }
  }
}

# Query for Recipe Collection for home (index) page
# getRecipeIndex

query recipeIndex($where: RecipeFilter) {
  __typename
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page
# getRecipePage

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    __typename
    items {
      ...RecipeDefault
    }
  }

  categoriesTaxonomy: taxonomyCollection(
    where: { slug: "categories" }
    order: title_ASC
    limit: 1
  ) {
    ...SchemaTaxonomy
  }

  cuisineTaxonomy: taxonomyCollection(
    where: { slug: "cuisine" }
    order: title_ASC
    limit: 1
  ) {
    ...SchemaTaxonomy
  }
}

query recipeEntryQuery($id: String!) {
  recipe(id: $id) {
    sys {
      id
      publishedVersion
    }

    ingredientsList
    instructionsList
  }
}

# used for recipe content structure migration
query recipeCollectionQuery {
  recipeCollection {
    total
    items {
      sys {
        id
        publishedVersion
      }

      ingredientsList
      instructionsList
    }
  }
}
