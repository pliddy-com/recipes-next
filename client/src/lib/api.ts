/* istanbul ignore file */

import { queryGraphQLContent } from './gqlClient';
// import * as fs from 'fs';

import {
  filterSlugs,
  filterTagsWithRecipes,
  filterTaxonomyItemsWithRecipes
} from './apiFilters';

import {
  NavMenuDataDocument,
  RecipeIndexDocument,
  RecipeIndexQueryVariables,
  RecipeListDocument,
  RecipeListQueryVariables,
  RecipePageDocument,
  RecipePageQueryVariables,
  RecipeSlugsDocument,
  RecipeSlugsQueryVariables,
  TagSlugsDocument,
  TagSlugsQueryVariables,
  NavMenuDataQuery,
  RecipeSlugsQuery,
  TagSlugsQuery,
  RecipeListQuery,
  LinkedEntriesQuery,
  LinkedEntriesDocument
} from '../types/queries';

import { hasValue } from './utils';

// const ACCESS_TOKEN = process.env.NEXT_PUBLIC_CONTENTFUL_MANAGEMENT_TOKEN ?? '';
// const SPACE_ID = process.env.NEXT_PUBLIC_CONTENTFUL_SPACE_ID ?? '';
// const API_ENDPOINT = process.env.NEXT_PUBLIC_CONTENTFUL_CONTENT_API;

// const client = contentful.createClient({
//   accessToken: ACCESS_TOKEN,
//   space: SPACE_ID
// });

// used to query content for nav menu taxonomy on client layout
export const getNavTaxonomy = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: NavMenuDataQuery = await queryGraphQLContent(
    NavMenuDataDocument
  );

  const { categories, cuisine, tags } = results;

  return {
    ...(categories && {
      categories: filterTaxonomyItemsWithRecipes({
        taxonomyCollection: categories
      })
    }),
    ...(cuisine && {
      cuisine: filterTaxonomyItemsWithRecipes({ taxonomyCollection: cuisine })
    }),
    ...(tags && {
      tags: filterTagsWithRecipes({ tagCollection: tags })
    })
  };
};

// used to get linked entries for recipe edit

export const getLinkedEntries = async () => {
  const { queryGraphQLContent } = await import('lib/gqlClient');

  const results: LinkedEntriesQuery = await queryGraphQLContent(
    LinkedEntriesDocument
  );

  const { images, tags } = results;

  return {
    ...(images && {
      images: images.items
    }),
    ...(tags && {
      tags: tags.items
    })
  };
};

// used by getStaticProps for recipe page
export const getRecipeSlugs = async (variables: RecipeSlugsQueryVariables) => {
  const results: RecipeSlugsQuery = await queryGraphQLContent(
    RecipeSlugsDocument,
    variables
  );

  const { recipeCollection } = results;

  return recipeCollection && recipeCollection.items
    ? filterSlugs(recipeCollection.items)
    : [];
};

// used by getStaticProps for tag page
export const getTagSlugs = async (variables?: TagSlugsQueryVariables) => {
  const results: TagSlugsQuery = await queryGraphQLContent(
    TagSlugsDocument,
    variables
  );

  const { tagCollection } = results;

  const tags =
    tagCollection &&
    filterTagsWithRecipes({
      tagCollection
    });

  return tags ? filterSlugs(tags) : [];
};

// used to query content for home index page
export const getRecipeIndex = async (variables?: RecipeIndexQueryVariables) => {
  const { recipeCollection } = await queryGraphQLContent(
    RecipeIndexDocument,
    variables
  );

  return recipeCollection ? recipeCollection.items.filter(hasValue) : [];
};

// used to query content for tag & category page
export const getRecipeList = async (variables?: RecipeListQueryVariables) => {
  const { tagCollection } = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  return tagCollection ? tagCollection.items.filter(hasValue) : [];
};

// used to query content for standalone recipe page
export const getRecipePage = async (variables?: RecipePageQueryVariables) => {
  const { recipeCollection, categoriesTaxonomy, cuisineTaxonomy } =
    await queryGraphQLContent(RecipePageDocument, variables);

  const categories =
    categoriesTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const cuisine =
    cuisineTaxonomy?.items.filter(hasValue)[0].childrenCollection?.items;

  const recipe = recipeCollection?.items.filter(hasValue)[0];

  return {
    ...(recipe && { recipe }),
    ...(categories && { categories }),
    ...(cuisine && { cuisine })
  };
};

// used to query content for tag index page
export const getTagIndex = async (variables?: RecipeListQueryVariables) => {
  const results: RecipeListQuery = await queryGraphQLContent(
    RecipeListDocument,
    variables
  );

  const { tagCollection } = results;

  return tagCollection
    ? filterTagsWithRecipes({
        tagCollection
      })
    : [];
};
