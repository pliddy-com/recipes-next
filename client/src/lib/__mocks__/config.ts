import { ConfigProps } from 'lib/config';

const microcopy = {
  index: {
    defaultTitle: 'Index Default Title',
    description: 'Index default description.'
  },
  notFound: {
    defaultTitle: 'Page Not Found',
    description: 'The requested url was not found'
  },
  recipe: {
    defaultDescription: 'Recipe default description for',
    defaultTitle: 'Recipe Default Title'
  },
  search: {
    defaultTitle: 'Results for',
    description: 'Search results'
  },
  site: {
    title: 'Site Title',
    domain: 'https://recipes.pliddy.com'
  },
  tag: {
    defaultDescription: 'Tag default description for',
    defaultTitle: 'Tag Default Title'
  },
  tagIndex: {
    defaultTitle: 'Tag Index Default Title',
    description: 'Tag index default description.'
  }
};

const config: ConfigProps = {
  microcopy
};

module.exports = config;
