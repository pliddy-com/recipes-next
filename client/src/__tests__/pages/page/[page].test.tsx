import '@testing-library/jest-dom';
import { act, render, waitFor } from '@testing-library/react';
import preloadAll from 'jest-next-dynamic';

import RecipeListPage, {
  getStaticPaths,
  getStaticProps
} from 'pages/page/[page]';

import { RecipeSummaryFragment } from 'types/queries';

import config from 'lib/config';
import * as api from 'lib/api';

jest.mock('lib/config');
jest.mock('lib/api');
jest.mock('components/PageHead/PageTags/PageTags');
jest.mock('layout/RecipeGridLayout/RecipeGridLayout');

const env = process.env;

describe('RecipeListPage in [page].tsx', () => {
  beforeEach(async () => {
    jest.resetModules();

    process.env = {
      ...env,
      NEXT_PUBLIC_SITE_URL: 'https://test.recipes.pliddy.com'
    };

    await preloadAll();
  });

  afterEach(() => {
    process.env = env;
  });

  describe('when there is content', () => {
    it('it renders the page', async () => {
      const getRecipeIndexSpy = jest.spyOn(api, 'getRecipeIndex');
      const recipeCollectionData = await api.getRecipeIndex();
      const page = 1;

      const expectedProps = {
        props: {
          pageContent: recipeCollectionData,
          preview: true,
          page
        },
        revalidate: 10
      };

      const context = {
        params: { page: '1' },
        preview: true
      };

      const expectedPaths = {
        fallback: false,
        paths: [{ params: { page: '1' } }]
      };

      const { asFragment, queryByTestId } = render(
        <RecipeListPage
          pageContent={recipeCollectionData as RecipeSummaryFragment[]}
          page={page}
          preview={false}
        />
      );

      // wait for dynamic component to load
      await act(async () => waitFor(() => queryByTestId('RecipeGrid')));

      // assert getStaticProps returns a value and manages preview default
      expect(await getStaticProps(context)).toEqual(expectedProps);
      expect(await getStaticPaths({})).toEqual(expectedPaths);

      expect(getRecipeIndexSpy).toHaveBeenCalledTimes(3);

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there is no page parameter', () => {
    it('it renders the page', async () => {
      // const getRecipeIndexSpy = jest.spyOn(api, 'getRecipeIndex');
      const recipeCollectionData = await api.getRecipeIndex();

      const expectedProps = {
        props: {
          pageContent: recipeCollectionData,
          preview: true,
          page: undefined
        },
        revalidate: 10
      };

      const context = {
        params: { page: undefined },
        preview: true
      };

      const { asFragment, queryByTestId } = render(
        <RecipeListPage
          pageContent={recipeCollectionData as RecipeSummaryFragment[]}
          page={undefined}
          preview={false}
        />
      );

      // wait for dynamic component to load
      await act(async () => waitFor(() => queryByTestId('RecipeGrid')));

      // assert getStaticProps returns a value and manages preview default
      expect(await getStaticProps(context)).toEqual(expectedProps);

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there is no page content', () => {
    it('it does not render the page', async () => {
      const { queryByTestId } = render(
        <RecipeListPage pageContent={[]} preview={false} page={undefined} />
      );

      // wait for dynamic component to load
      await act(async () =>
        waitFor(() => expect(queryByTestId('RecipeGrid')).toBeNull())
      );
    });
  });

  describe('when there is no config object', () => {
    // before each test, delete microcopy node from config
    beforeEach(() => {
      delete config.microcopy;
    });

    it('it does not render the page', async () => {
      const { queryByTestId } = render(
        <RecipeListPage pageContent={[]} preview={false} page={undefined} />
      );

      // wait for dynamic component to load
      await act(async () =>
        waitFor(() => expect(queryByTestId('RecipeGrid')).toBeNull())
      );
    });
  });
});
