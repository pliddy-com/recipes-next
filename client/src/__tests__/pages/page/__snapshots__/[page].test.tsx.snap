// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeListPage in [page].tsx when there is content it renders the page 1`] = `
<DocumentFragment>
  <title>
    Index Default Title (Page 1)
  </title>
  <meta
    content="Index default description."
    name="description"
  />
  <script
    data-testid="recipe-list-schema"
    id="recipe-list-schema"
    type="application/ld+json"
  >
    {"@context":"https://schema.org","@type":"ItemList","name":"Index Default Title (Page 1)","description":"Index default description.","url":"https://test.recipes.pliddy.com","numberOfItems":2,"itemListElement":[{"@type":"ListItem","position":1,"name":"Title 1","url":"https://test.recipes.pliddy.com/recipe/title-1"},{"@type":"ListItem","position":2,"name":"Title 2","url":"https://test.recipes.pliddy.com/recipe/title-2"}]}
  </script>
</DocumentFragment>
`;

exports[`RecipeListPage in [page].tsx when there is no page parameter it renders the page 1`] = `
<DocumentFragment>
  <title>
    Index Default Title (Page undefined)
  </title>
  <meta
    content="Index default description."
    name="description"
  />
  <script
    data-testid="recipe-list-schema"
    id="recipe-list-schema"
    type="application/ld+json"
  >
    {"@context":"https://schema.org","@type":"ItemList","name":"Index Default Title (Page undefined)","description":"Index default description.","url":"https://test.recipes.pliddy.com","numberOfItems":2,"itemListElement":[{"@type":"ListItem","position":1,"name":"Title 1","url":"https://test.recipes.pliddy.com/recipe/title-1"},{"@type":"ListItem","position":2,"name":"Title 2","url":"https://test.recipes.pliddy.com/recipe/title-2"}]}
  </script>
</DocumentFragment>
`;
